Add transfer benchmark script

JordonPhillips opened the pull request:
This adds a script that benchmarks all the cli transfer commands.
The commands are interleaved such that no additional setup is
necessary between command runs.

The weakness of this script is that it can't run individual commands,
though I think if we want that it would make sense as additional
scripts.

I also currently have it set to read from json defintions for cases,
but it may be wiser to restrict to a single test case per run and have
the additional arguments passed on the command line.

cc @jamesls @kyleknap

jamesls commented:
It would be nice to improve the output to differentiate commands we're running vs. output from the tool.  I ran into an error trying out this command (not sure what), and the output makes it a little hard to track down:

```
aws s3 sync s3://bucket-name /var/folders/rb/4c1d0xbd18g_wxjcjlvb0whn1cjh5m/T/tmpDeFvAi
aws s3 cp s3://bucket-name/largefile /var/folders/rb/4c1d0xbd18g_wxjcjlvb0whn1cjh5m/T/tmpDeFvAi/file
download: s3://bucket-name/largefile to ../../../../var/folders/rb/4c1d0xbd18g_wxjcjlvb0whn1cjh5m/T/tmpDeFvAi/file
aws s3 mv /var/folders/rb/4c1d0xbd18g_wxjcjlvb0whn1cjh5m/T/tmpDeFvAi/file s3://bucket-name/largefile
move: ../../../../var/folders/rb/4c1d0xbd18g_wxjcjlvb0whn1cjh5m/T/tmpDeFvAi/file to s3://bucket-name/largefile
aws s3 sync s3://bucket-name /var/folders/rb/4c1d0xbd18g_wxjcjlvb0whn1cjh5m/T/tmpDeFvAi
aws s3 cp s3://bucket-name/largefile /var/folders/rb/4c1d0xbd18g_wxjcjlvb0whn1cjh5m/T/tmpDeFvAi/file
download: s3://bucket-name/largefile to ../../../../var/folders/rb/4c1d0xbd18g_wxjcjlvb0whn1cjh5m/T/tmpDeFvAi/file
aws s3 mv /var/folders/rb/4c1d0xbd18g_wxjcjlvb0whn1cjh5m/T/tmpDeFvAi/file s3://bucket-name/largefile
move: ../../../../var/folders/rb/4c1d0xbd18g_wxjcjlvb0whn1cjh5m/T/tmpDeFvAi/file to s3://bucket-name/largefile
aws s3 sync s3://bucket-name /var/folders/rb/4c1d0xbd18g_wxjcjlvb0whn1cjh5m/T/tmpDeFvAi
aws s3 cp /var/folders/rb/4c1d0xbd18g_wxjcjlvb0whn1cjh5m/T/tmpDeFvAi/file s3://bucket-name/largefile2

The user-provided path /var/folders/rb/4c1d0xbd18g_wxjcjlvb0whn1cjh5m/T/tmpDeFvAi/file does not exist.
aws s3 mv s3://bucket-name/largefile2 /var/folders/rb/4c1d0xbd18g_wxjcjlvb0whn1cjh5m/T/tmpDeFvAi/file
An error occurred (404) when calling the HeadObject operation: Key "largefile2" does not exist
Completed 1 part(s) with ... file(s) remaining
aws s3 sync /var/folders/rb/4c1d0xbd18g_wxjcjlvb0whn1cjh5m/T/tmpDeFvAi s3://bucket-name
aws s3 rm s3://bucket-name/largefile2
An error occurred (404) when calling the HeadObject operation: Key "largefile2" does not exist
Completed 1 part(s) with ... file(s) remaining
aws s3 cp /var/folders/rb/4c1d0xbd18g_wxjcjlvb0whn1cjh5m/T/tmpDeFvAi/file s3://bucket-name/largefile2

The user-provided path /var/folders/rb/4c1d0xbd18g_wxjcjlvb0whn1cjh5m/T/tmpDeFvAi/file does not exist.
aws s3 mv s3://bucket-name/largefile2 /var/folders/rb/4c1d0xbd18g_wxjcjlvb0whn1cjh5m/T/tmpDeFvAi/file
An error occurred (404) when calling the HeadObject operation: Key "largefile2" does not exist
Completed 1 part(s) with ... file(s) remaining
aws s3 sync /var/folders/rb/4c1d0xbd18g_wxjcjlvb0whn1cjh5m/T/tmpDeFvAi s3://bucket-name
aws s3 rm s3://bucket-name/largefile2
An error occurred (404) when calling the HeadObject operation: Key "largefile2" does not exist
Completed 1 part(s) with ... file(s) remaining
aws s3 cp /var/folders/rb/4c1d0xbd18g_wxjcjlvb0whn1cjh5m/T/tmpDeFvAi/file s3://bucket-name/largefile2

```

jamesls commented on scripts/performance/benchmark-s3-transfer:
```diff
@@ -0,0 +1,243 @@
+    if script is None:
+        script = _get_benchmark_script()
+
+    if script is None:
```
You could just fold all this into the `required=True`, and a custom `type=your_validator` in the `add_argument` call for `--benchmark-script`.

jamesls commented on scripts/performance/benchmark-s3-transfer:
```diff
@@ -0,0 +1,243 @@
+        help=('The benchmark script to run the commands with. This should be '
+              'from s3transfer.'))
+    parser.add_argument(
+        '-o', '--result-dir', default='results',
```
This arg isn't honored.  Took me a while to figure out why I wasn't getting any results.

jamesls commented:
I keep getting the above error when I try this out.  I've tried several configurations (same bucket, different bucket), and it always seems to happen midway through.

I think I would prefer to make this more granular.  Perhaps the test case json files can specify which type of transfers they want tested so if you just want a small subset to test you don't have to test every single command per invocation.  It would also make troubleshooting easier.

JordonPhillips commented:
Looks like that error is due to the destination key being different from the source, which messes with the interleaving. I'll update it so only a destination bucket is provided.

jamesls commented on scripts/performance/benchmark-cp:
```diff
@@ -0,0 +1,93 @@
+        summarize(args.summarize_script, args.result_dir)
+
+        if not args.no_cleanup:
+            clean(args.destination, args.recursive)
```
I think this should be wrapped in a `try/finally`.  I think we'd want to clean up even in the event that the bench mark scripts (or summarize scripts) fail.

jamesls commented on scripts/performance/benchmark_utils.py:
```diff
@@ -0,0 +1,119 @@
+        shutil.copy(source, destination)
+
+
+def clean(destination, recursive):
```
I think we need to be careful about this.  Especially in the `--recursive` case, if I had specified my `~/` directory as the dest and I was downloading objects from S3 that were broken down by prefix:

```
s3://bucket/foo
s3://bucket/bar
```

Then this script (with a source of `s3://bucket/`) would download to `~/foo` and `~/bar` if I was testing `s3->local`, but as part of cleanup it would `rm -rf` my entire home directory.

I think we should try to be more conservative in how my cleanup.  The ideal result is we only remove what we downloaded/uploaded, but I realize we would need to track that information somehow.  Is there a safer way we can clean up?

JordonPhillips replied:
We could create our own directory within the specified directory to isolate things, but that defeats half of the point of being able to specify your own location. We could list all the objects that would be copied, build those file paths, and then delete them. It would take much longer, but it would be safer.

jamesls commented on scripts/performance/benchmark-mv:
```diff
@@ -0,0 +1,98 @@
+
+        if not args.no_cleanup:
+            clean(args.destination, args.recursive)
+        clean(backup_path, args.recursive)
```
This applies in general to all the calls to `clean()`.  I think we should make a best effort to cleanup even in the case of errors.

jamesls commented on scripts/performance/benchmark-mv:
```diff
@@ -0,0 +1,98 @@
+    
+
+def main():
+    parser = argparse.ArgumentParser()
```
This looks like an exact duplicate in the other `benchmark-*` scripts.  Seems like a candidate for moving into `benchmark_utils`.

jamesls commented:
Overall seems to be working for me.  My only high level comment would be to move more stuff over to `benchmark_utils` to cut down on the duplication between scripts.

kyleknap commented:
As a heads up before I start commenting, I am trying your scripts out right now and I am going to make comments as I run into things so if I do not comment for a while there is a chance that I have more. I will let you know when I have no more comments.

kyleknap commented on scripts/performance/benchmark_utils.py:
```diff
@@ -0,0 +1,119 @@
+        return int(value[:-len(suffix)]) * multiplier
+
+
+def summarize(script, directory, filename='summary.txt'):
```
I am not sure that I am a fan of adding the ``summarize.txt`` file because if I ever want to run ``summarize`` on that directory again, I would not be able to by doing ``/summarize my-results/*`` assuming my-results is the output directory. I would prefer the directory is reserved to the CSV's.

JordonPhillips commented:
This is ready to look at again. I've addressed all the comments and also removed the monolithic benchmark script since its marginal utility is minuscule with the individual scripts in place.

kyleknap commented on scripts/performance/benchmark_utils.py:
```diff
@@ -0,0 +1,176 @@
+
+
+def backup(source, recursive):
+    if source[:5] == 's3://':
```
This looks similar to ``get_random_subfolder``. Could we reuse that at all?

kyleknap commented on scripts/performance/benchmark_utils.py:
```diff
@@ -0,0 +1,176 @@
+}
+
+
+def human_readable_to_bytes(value):
```
This function is no longer used.

kyleknap commented:
To capture the offline discussion, I think we should consolidate a lot of this logic (especially when you iterate over and do the runs) into more general functions as it will allow us to easily update these scripts if we want to do something like add more args or fix a bug.

JordonPhillips commented:
Updated. Not 100% on the cleanup / upkeep bits, what do y'all think?

jamesls commented on scripts/performance/benchmark_utils.py:
```diff
@@ -0,0 +1,288 @@
+
+
+def get_random_subfolder(destination):
+    """Get a random subdirectory of a given directory."""
```
Might be worth clarifying/renaming this to make it clear we will *create* a new random subfolder under the provided destination for local files (vs. randomly picking an existing subdirectory).

jamesls commented:
:shipit:  Looks good to me.  I've been trying this out locally and it's been working great.

kyleknap commented:
Nice that cleaned it up a good amount. Looks good. I think ``human_readable_to_bytes`` is still around so you can delete that. But otherwise :ship:
